[FIX JENKINS-38391] Show admin monitors on most URLs

daniel-beck opened the pull request:
Building on the preparatory work in #2546, this change makes admin monitors show up on most URLs (for admins).

My thanks to @kzantow who made this look like an adult designed it rather than a five year old.

---

> ![screen shot closed](https://cloud.githubusercontent.com/assets/1831569/18694289/0bed9236-7fa9-11e6-937c-67259c04865f.png)

---

> ![screen shot open](https://cloud.githubusercontent.com/assets/1831569/18694293/109f54ae-7fa9-11e6-9697-f44b8ef5f8d6.png)

---

Keith and I disagreed about whether something in core should be using the `PageDecorator` extension point approach. I prefer this approach over bloating `/lib/layout/layout.jelly` further. If there are actual issues with this approach, those should be addressed in the `PageDecorator` extension point (maybe by adding a `head.jelly` include to extend the search box/login link/… area).

Known issues:
- Reverse proxy monitor does not work when shown on a different URL than `/manage`, it will always complain. So that one is excluded for now.
- Admin monitors external to core may not expect being shown on a URL different from `/manage` (or rather a URL resulting in different relative link targets), but there's little we can (or should) do about that here.

Despite these, I do not consider this work in progress unless someone objects.

CC @rtyler who wanted me to get this done.

@reviewbybees @jenkinsci/code-reviewers

daniel-beck committed:
[FIX JENKINS-38391] Show admin monitors on most URLs

daniel-beck commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/footer.jelly:
```diff
@@ -0,0 +1,159 @@
+                    </div>
+                </j:forEach>
+                <p style="text-align: center; margin: 10px 0 0 0;">
+                    <a href="${rootURL}/manage" onclick="document.location.href='${rootURL}/manage';">${%Manage Jenkins}</a>
```
Can anyone guess what the purpose of this may have been? I have no idea what I thought when I did that.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

daniel-beck committed:
[JENKINS-38391] Further admin monitor refinements

- Add tooltip to notification button
- Move CSS/JS into an adjunct
- Check adcitve admin monitors status last
- Don't show on /configure URL

oleg-nenashev commented on core/src/main/java/jenkins/management/AdministrativeMonitorsDecorator.java:
```diff
@@ -0,0 +1,136 @@
+            return false;
+        }
+
+        StaplerRequest req = Stapler.getCurrentRequest();
```
NIT: could be checked first if it is really required

oleg-nenashev commented on core/src/main/java/jenkins/management/AdministrativeMonitorsDecorator.java:
```diff
@@ -0,0 +1,136 @@
+     * @throws IOException
+     * @throws ServletException
+     */
+    public boolean shouldDisplay() throws IOException, ServletException {
```
Should it be Restricted?

daniel-beck replied:
The entire class is restricted.

oleg-nenashev commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/footer.jelly:
```diff
@@ -0,0 +1,159 @@
+-->
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+    <j:if test="${it.shouldDisplay()}">
```
From what I see this call immediately causes calculation of administration monitors in a synchronous mode on **every** page. I would rather 🐛 it, because monitors have no calculation timeout for <code>isActivated()</code>. If a monitor goes crazy, it may cause complete DoS of the web interface

oleg-nenashev reviewed (changes requested):
🐛 from me, because this PR leads to the calculation of AdministrativeMonitor statuses for admins during page rendering. If something goes wrong with AdministrativeMonitors, it will prevent admins from accessing Jenkins WebUI, so it will be difficult to disable broken monitors.

IMHO calculation of <code>isActivated</code> should be moved to AJAX request, timeout is maybe also required

daniel-beck commented:
> If something goes wrong with AdministrativeMonitors, it will prevent admins from accessing Jenkins WebUI, so it will be difficult to disable broken monitors.

@oleg-nenashev Given that literally everything else is also synchronously determined, including the parts updating via AJAX, I don't really understand this concern. But should be sufficient to exclude `/configure` which allows disabling admin monitors in the unlikely case this ever happens to restore UI functionality. WDYT?

daniel-beck commented:
@oleg-nenashev Also, has there ever been an admin monitor where this actually was a problem?

rtyler commented:
I really like how this looks, I'm excited to see this land and eventually get into the LTSes :)

oleg-nenashev commented:
@daniel-beck 

> But should be sufficient to exclude /configure which allows disabling admin monitors in the unlikely case this ever happens to restore UI functionality. WDYT?

Also "login" and "manage" pages. But then it should be fine

oleg-nenashev commented on core/src/main/java/jenkins/management/AdministrativeMonitorsDecorator.java:
```diff
@@ -0,0 +1,136 @@
+                // TODO make reverse proxy monitor work when shown on any URL
+                continue;
+            }
+            if (am.isEnabled() && am.isActivated()) {
```
Nice2have: Catch runtime exceptions (just in case somebody provides unstable extension point implementation)

daniel-beck replied:
Then it will just evaluate to false and not show up. NBD.

daniel-beck commented:
> Also "login" and "manage" pages. But then it should be fine

It's in the footer, so it executes last during rendering. It's a problem for JS driven config forms, but shouldn't be for these pages.

recena commented:
🐛  and 👎  I don't like the UX and the front-end code. More CSS and JS embedded.

If we really want a status bar where show information, go forward, but these changes only do more fragile improvements in the UI in short term.

daniel-beck commented:
@recena In what way is this fragile? Also, how would you separate the CSS/JS? Stapler adjuncts?

Please explain how I can move this forward.

kzantow commented:
@recena I don't think you expanded on the :bug: properly, there is no clear way to resolve it.

recena commented:
@daniel-beck If this is something provided by core, why the JS and CSS are not provided the same way than other common JS and CSS resources?

Fragile? Let me re-design the header and you will see if the current implementation works.

daniel-beck commented:
@recena 

> If this is something provided by core, why the JS and CSS are not provided the same way than other common JS and CSS resources?

Again, tell me what you want me to change and I will do so.

> Fragile? Let me re-design the header and you will see if the current implementation works.

No different from anything else really, and as it's in the same component (just not in the same source file), it's easy enough to adapt as needed.

recena commented:
🐝 and 👍 but this is need a follow-up for a more general solution if we want really implementing a toolbar for status info.

daniel-beck commented:
@oleg-nenashev Please re-review. I addressed your comments while keeping the synchronous behavior. Admins can now access `/configure` in the unexpected case a monitor goes crazy.

aheritier commented:
👍 I love the idea but I don't have the skills to judge about UI comments

oleg-nenashev commented:
Does the change also filters away "/computer/configure" ? Monitors are actually being managed on this page

daniel-beck commented:
@oleg-nenashev No, only `/configure`. The naming `ignoredJenkinsRestOfUrls` is deliberate, this only applies to URLs served by `Jenkins` (and not e.g. `ComputerSet`).

oleg-nenashev commented:
🐝 since https://github.com/jenkinsci/jenkins/pull/2546 allows solving the issue on the /configure page

scherler commented:
I just stumbled on the appearance of this "feature" @daniel-beck IMHO that is one of the most confusing things I saw. 

Suddenly I have a big red box with a number on it (before the search box) there are the following problems:
- position of notification are normally aligned to the right
- notifications are marked/indicated by an icon not a box that suddenly appears out of no where. This means that you have 2 icons one for "no-notifications" and another for notifications (as we all know from our mobil phones)
- the css you are using are done for a certain display size, if you do not have the size see the screenshot on what happens
![screenshot from 2016-10-18 16-58-47](https://cloud.githubusercontent.com/assets/596701/19483346/4f0bdc10-9554-11e6-8ad0-ea96af9d6c50.png)

Without to step on anybodies toes I agree with @recena when he called it a hacky solution. IMO that is not helping the UX but the opposite. 

I know it is merged but :bug: from me because of all the above

oleg-nenashev commented:
@scherler Well, it's a good opportunity to create a new PR ;)

scherler commented:
@oleg-nenashev I may actually be bothered enough to do it ;)

daniel-beck commented:
@scherler 

> the css you are using are done for a certain display size, if you do not have the size see the screenshot on what happens

Uh… you mean the screenshot in which the problem clearly is the absolutely positioned 'Try Blue Ocean' button as is obvious through the simple fact that it overlaps the search bar as well?

Here's Jenkins 2.17, with the exact same problem:

![](https://cloud.githubusercontent.com/assets/1831569/19484490/18893e7c-9558-11e6-8807-e661cd85e485.png)

Clearly, this is a BlueOcean problem.

> notifications are marked/indicated by an icon not a box that suddenly appears out of nowhere. This means that you have 2 icons one for "no-notifications" and another for notifications (as we all know from our mobil phones)

Not clear what you mean. Are you saying we should show something (e.g. a green "0") when there's no admin monitor active, and have a popup with "no messages"? That would be trivial to do. But note that we cannot show 'read notifications' because admin monitors currently don't work like that.

>  IMO that is not helping the UX but the opposite. 

To clarify, you'd rather _not_ inform users about severe problems with their Jenkins instance instead of doing it in a way you consider not aesthetically pleasing?

scherler commented:
@daniel-beck yes the try-blue-ocean button is as well a hack, but if you in "blueocean" you will not see it ;) Jokes apart that button should be fixed as well I agree.

Regarding the notifications (position and appearance) here to screen shots from github (how it could look like).

![screenshot from 2016-10-18 17-36-38](https://cloud.githubusercontent.com/assets/596701/19484996/e76ceb48-9559-11e6-9b6b-94e0acd9d0cd.png)
![screenshot from 2016-10-18 17-36-12](https://cloud.githubusercontent.com/assets/596701/19484995/e75cbca0-9559-11e6-9eb1-0b3b978a9b8b.png)

The point being that the design with and without the notifications are aligned the same. If there are no notification just do not link to a popup, but still have the icon. 

I prefer that we have a well designed notification bar, since I doubt that all notification are "red" -> critical to begin with. Under a UX perspective, to be honest, yes it would be better not to show it rather then show it at all. It provokes confusion on what it means and where it is placed.
